code/Controllers/CMSMain.php

```diff
@@ -1038,56 +1038,93 @@ public function getBreadcrumbsBackLink()
             ->Link;
     }
 
+    /**
+     * Generate a breadcrumb for based on the page currently being viewed.
+     * @param bool $unlinked Wheter to include links
+     * @return ArrayList<ArrayData>
+     */
     public function Breadcrumbs($unlinked = false)
     {
         $items = new ArrayList();
 
-        if ($this->TreeIsFiltered()) {
-            $items->push(new ArrayData([
-                'Title' => CMSPagesController::menu_title(),
-                'Link' => ($unlinked) ? false : $this->LinkPages()
-            ]));
-            $items->push(new ArrayData([
-                'Title' => _t('SilverStripe\\CMS\\Controllers\\CMSMain.SEARCHRESULTS', 'Search results'),
-                'Link' => ($unlinked) ? false : $this->LinkPages()
-            ]));
-
-            $this->extend('updateBreadcrumbs', $items);
+        if (($this->getAction() !== 'index') && ($record = $this->currentPage())) {
+            // The page is being edited
+            $this->buildEditFormBreadcrumb($items, $record, $unlinked);
+        } else {
+            // All other scenarios include a "Pages" first crumb
+            $this->pushCrumb(
+                $items,
+                CMSPagesController::menu_title(),
+                $unlinked ? false : $this->LinkPages()
+            );
 
-            return $items;
+            if ($this->TreeIsFiltered()) {
+                // Showing search results
+                $this->pushCrumb(
+                    $items,
+                    _t(CMSMain::class . '.SEARCHRESULTS', 'Search results'),
+                    ($unlinked) ? false : $this->LinkPages()
+                );
+            } elseif ($parentID = $this->getRequest()->getVar('ParentID')) {
+                // We're navigating the listview. ParentID is the page whose
+                // children are currently displayed.
+                if ($page = SiteTree::get()->byID($parentID)) {
+                    $this->buildListViewBreadcrumb($items, $page);
+                }
+            }
         }
 
-        // Check if we are editing a page
-        /** @var SiteTree $record */
-        $record = $this->currentPage();
-        if (!$record) {
-            $items->push(new ArrayData([
-                'Title' => CMSPagesController::menu_title(),
-                'Link' => ($unlinked) ? false : $this->LinkPages()
-            ]));
+        $this->extend('updateBreadcrumbs', $items);
```

Comment:
The original implementation would build the breadcrumb for a specific scenario, call `updateBreadcrumbs` and return.

Seems more logical to build to use `if` statements to make sure only one breadcrumb is built and then call `updateBreadcrumbs` once.